### PR TITLE
Added support for pandas 2

### LINKED ISSUE #4252: Add support for pandas 2.0 to EvalML

### PR DESCRIPTION
**Pull request by christopherbunn**
Resolves #4252

### CODE REVIEW

**Comment by machineFL** [comment]
No linked issues found. Please add the corresponding issues in the pull request description. <br/> [Use GitHub automation to close the issue when a PR is merged](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)

**Comment by gsheni** [comment]
@christopherbunn Woodwork and Featuretools now support pandas 2.0

**Review by christopherbunn — COMMENTED** [review]
Few points to consider:

File: evalml/pipelines/utils.py
Comment (eccabay) [inline evalml/pipelines/utils.py:1]: Looks like your rebase backed out too much! 😱

File: evalml/pipelines/components/transformers/preprocessing/decomposer.py [hunk 1]
```diff
@@ -151,7 +151,9 @@ def determine_periodicity(
                 period is detected, returns None.
 
         """
-        X, y = cls._handle_nullable_types(cls, X, y)
+        # Only need to handle nullable types on pandas < 2. Kept for backwards compatibility with pandas 1.x.
+        if pd.__version__[0] == "1":
```
Comment (eccabay) [inline evalml/pipelines/components/transformers/preprocessing/decomposer.py:155]: Is this really the best way to check the version? It looks janky and brittle

File: evalml/tests/conftest.py
Comment (eccabay) [inline evalml/tests/conftest.py:1]: More code that shouldn't be deleted!

File: evalml/tests/data_checks_tests/test_datetime_format_data_check.py [hunk 2]
```diff
@@ -244,7 +244,7 @@ def test_datetime_format_data_check_multiple_errors():
         + ["2021-01-31", "2021-02-02", "2021-02-04"]
         + pd.date_range("2021-02-05", periods=90).tolist()
     )
-    X = pd.DataFrame({"dates": dates})
+    X = pd.DataFrame({"dates": dates}, dtype="datetime64[ns]")
```
Comment (eccabay) [inline evalml/tests/data_checks_tests/test_datetime_format_data_check.py:247]: Why are all these necessary?

File: evalml/tests/pipeline_tests/test_pipeline_utils.py
Comment (eccabay) [inline evalml/tests/pipeline_tests/test_pipeline_utils.py:1]: More deleted code

File: pyproject.toml [hunk 3]
```diff
@@ -30,7 +30,7 @@ license = {file = "LICENSE"}
 requires-python = ">=3.8,<4"
 dependencies = [
     "numpy >= 1.21.0",
-    "pandas >= 1.5.0, <2.0.0",
+    "pandas >= 2.0.2",
```
Comment (eccabay) [inline pyproject.toml:33]: Again, we need to figure out the correct versioning 🤔

**Review by eccabay — CHANGES_REQUESTED** [review]
Thanks for doing this! We need to make sure there isn't any accidentally deleted code and figure out what the minimum version should be before this can go in.

File: .github/meta.yaml [hunk 4]
```diff
@@ -25,7 +25,7 @@ outputs:
         - setuptools ==58.0.4
       run:
         - numpy >=1.21.0
-        - pandas >=1.5.0, <2.0.0
+        - pandas >=2.0.2
```
[the remark this answers is not in the document]
Comment (christopherbunn) [reply .github/meta.yaml]: Oops, meant to just uncap.

File: core-requirements.txt [hunk 5]
```diff
@@ -1,5 +1,5 @@
 numpy>=1.21.0
-pandas>=1.5.0, <2.0.0
+pandas>=2.0.3
```
[the remark this answers is not in the document]
Comment (christopherbunn) [reply core-requirements.txt]: Reverted and uncapped.

File: docs/source/release_notes.rst [hunk 6]
```diff
@@ -3,8 +3,8 @@ Release Notes
 **Future Releases**
     * Enhancements
         * Updated regression metrics to handle multioutput dataframes as well as single output series :pr:`4233`
-        * Added stacking and unstacking utility functions to work with multiseries data :pr:`4250`
```
[the remark this answers is not in the document]
Comment (christopherbunn) [reply docs/source/release_notes.rst:6]: Placed this back in 😅

File: evalml/pipelines/utils.py
Comment (christopherbunn) [reply evalml/pipelines/utils.py:1]: My bad, added back in

File: evalml/tests/pipeline_tests/test_pipeline_utils.py
Comment (christopherbunn) [reply evalml/tests/pipeline_tests/test_pipeline_utils.py:1]: Reverted

File: evalml/tests/conftest.py
Comment (christopherbunn) [reply evalml/tests/conftest.py:1]: Reverted

File: evalml/pipelines/components/transformers/preprocessing/decomposer.py (hunk 1 above)
Comment (christopherbunn) [reply evalml/pipelines/components/transformers/preprocessing/decomposer.py:155]: Yeah I'm not a huge fan of this too to be honest. [This](https://stackoverflow.com/questions/20612645/how-to-find-the-installed-pandas-version) made it seem like this was the best option, given that `pd.__version__` returns the version number as a string. Less pandas-specific resources [here](https://stackoverflow.com/questions/710609/checking-a-python-module-version-at-runtime) still return the version as as string. I am very much open to alternative implementations/suggestions if anyone has any!

File: pyproject.toml (hunk 3 above)
Comment (christopherbunn) [reply pyproject.toml]: Reverted to previous but uncapped.

File: evalml/tests/data_checks_tests/test_datetime_format_data_check.py (hunk 2 above)
Comment (christopherbunn) [reply evalml/tests/data_checks_tests/test_datetime_format_data_check.py:247]: The previous behavior left it as object types, which in pandas 2 raises an error in [this line](https://github.com/alteryx/woodwork/blob/732cf90f051cef0d65fa3a3ecfefecf9d07743b2/woodwork/statistics_utils/frequency_inference/_determine_duplicate_values.py#L25) of woodwork. By casting it as an explicit datetime object, we're able to still utilize this function.

**Review by christopherbunn — COMMENTED** [review]
My bad, most of the code regressions are due to some bad merge work with your PR @eccabay 😅

File: evalml/pipelines/components/transformers/preprocessing/decomposer.py (hunk 1 above)
Comment (eccabay) [reply evalml/pipelines/components/transformers/preprocessing/decomposer.py:155]: Would it make more sense to do something like `int(pd.__version__.split('.')[0]) < 2`? I highly doubt this code would be here if pandas ever got to a version 10.x, but it feels a little safer than string comparison?

**Review by eccabay — APPROVED** [review]

**Review by jeremyliweishih — APPROVED** [review]
LGTM but +1 on @eccabay 's comments re checking for version

File: evalml/pipelines/components/transformers/preprocessing/decomposer.py (hunk 1 above)
Comment (christopherbunn) [reply evalml/pipelines/components/transformers/preprocessing/decomposer.py:155]: Good call, will replace.

**Pull request merged**